std/process.d

```diff
@@ -2642,15 +2642,45 @@ class ProcessException : Exception
     version (Windows)
     static ProcessException newFromLastError(string customMsg = null,
                                              string file = __FILE__,
-                                             size_t line = __LINE__)
+                                             size_t line = __LINE__,
+                                             string lastMsg = sysErrorString(GetLastError()))
     {
-        auto lastMsg = sysErrorString(GetLastError());
         auto msg = customMsg.empty ? lastMsg
                                    : customMsg ~ " (" ~ lastMsg ~ ')';
         return new ProcessException(msg, file, line);
     }
+    // ditto
+    version (Windows)
+    static ProcessException newCreateProcessError(const(char)[] commandLine,
+                                                  string file = __FILE__,
+                                                  size_t line = __LINE__)
+    {
+        const customMsg = "Failed to spawn new process";
+        const lastError = GetLastError();
+        string lastMsg;
+        if (lastError == ERROR_FILE_NOT_FOUND)
+            lastMsg = "Cannot find file '" ~ cast(string)peelApplicationName(commandLine) ~ "'";
+        else
+            lastMsg = sysErrorString(lastError);
+        return newFromLastError(customMsg, file, line, lastMsg);
+    }
 }
 
+// Peel the application name from a windows command-line string meant for CreateProcess
+version (Windows)
+private auto peelApplicationName(inout(char)[] commandLine)
+{
+    import std.string : indexOf;
+    if (commandLine.length == 0) return commandLine;
+    char endChar = ' ';
+    if (commandLine[0] == '"')
+    {
+        commandLine = commandLine[1..$];
+        endChar = '"';
+    }
+    const end = commandLine.indexOf(endChar);
+    return commandLine[0 .. (end == -1) ? $ : end];
+}
```

Comment:
Why is this function at all necessary? `std.process` functions already take the command as an array of arguments, and its first element will be the executable name. This function isn't even correct, it would have to call or precisely replicate the logic of `CommandLineToArgvW`, which it doesn't.

Reply:
`spawnProcessImpl` actually takes a single string, since that's what `CreateProcessW` takes.  And I think the logic is correct here.  Take a look at the documentation for `CreateProcess`.  It finds the application name based on the first whitespace characters, unless it starts with a `"` character.  And windows filenames cannot contain quotes so you don't need to handle escapes in that instance.

https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw

Reply:
Here's the applicable documentation on it:

> The lpApplicationName parameter can be NULL. In that case, the module name must be the first white space–delimited token in the lpCommandLine string. If you are using a long file name that contains a space, use quoted strings to indicate where the file name ends and the arguments begin....

Reply:
> `spawnProcessImpl` actually takes a single string

This is a very bad reason to 1) undo the work that's done in the same module, and 2) duplicate the work of the operating system.

>  And I think the logic is correct here.

No. Sorry, it's not even close. For one thing, Windows ignores whitespace at the end of filenames. For another, tabs can also be used to separate arguments. There are likely many other subtle differences in behavior which would be pointless to try to match.

Sorry, the way you are approaching this is not good at all. Please rework this submission.